lib/experimental/cmdline.nim

```diff
@@ -0,0 +1,2856 @@
+                               ## be `some(string)` for flags with non-optional
+                               ## value.
+    flag*: FlagId ## Handle to the flag.
+
+  PositionalError* = object of ParseError
+    ## An error parsing positional parameters.
+    positionalValue*: string ## Input value causing the error
+  UnknownPositionalError* = object of PositionalError
+    ## The positional parsed was not recognized.
+  MissingPositionalError* = object of PositionalError
+    ## A required positional parameter is missing from input.
+    positional*: PositionalId ## Handle to the positional
+  InvalidPositionalError* = object of PositionalError
+    ## The positional parsed was invalid.
+    ##
+    ## The `ValueError` causing this can be found in the `parent` field.
+    positional*: PositionalId ## Handle to the positional
+
+  CommandError* = object of ParseError
+    ## An error parsing command.
+    commandName*: string ## Input value causing the error.
+  UnknownCommandError* = object of CommandError
+    ## The command parsed was not recognized.
+  MissingCommandError* = object of CommandError
+    ## A required command is missing from input.
+  InvalidCommandError* = object of CommandError
+    ## The command parsed was rejected by parser.
+    ##
+    ## The `ValueError` causing this can be found in the `parent` field.
+    targetCommand*: CommandId ## Handle to the target command
+
+  HelpError* = object of ParseError
+    ## Help was requested using `Action.ShowHelp`.
+    paramName*: string ## Name of the parameter triggering help, as specified by
+                       ## input.
+    param*: ParameterId ## Handle to the parameter triggering help.
+
+  ParameterKind* {.pure.} = enum
+    Command
+    Flag
+    Positional
+
+  ParameterId* = distinct uint32
+    ## A command line parameter. Values of this type are tied to the originating
+    ## `Cli` instance.
+    ##
+    ## A `ParameterId` might be a `CommandId`, `FlagId`, or `PositionalId`. The
+    ## `classify` function can be used to distinguish between them.
+
+  CommandId* = distinct ParameterId
+    ## A command line subcommand. Values of this type are tied to the
+    ## originating `Cli` instance.
+
+  FlagId* = distinct ParameterId
+    ## A command line flag. Values of this type are tied to the originating
+    ## `Cli` instance.
+
+  PositionalId* = distinct ParameterId
+    ## A command line positional parameter. Values of this type are tied to the
+    ## originating `Cli` instance.
+
+const
+  InvalidParameter = high(ParameterId)
+    ## An invalid parameter
+  RootCommand* = CommandId(0)
+    ## The top-level command of `Cli`
+
+proc hash(x: ParameterId): Hash {.borrow.}
+
+proc `==`*(a, b: ParameterId): bool {.borrow.}
+proc `==`*(a, b: CommandId): bool {.borrow.}
+proc `==`*(a, b: FlagId): bool {.borrow.}
+proc `==`*(a, b: PositionalId): bool {.borrow.}
+
+# FIXME: Put this in system.nim
+func drop[T](_: sink T) = discard
```

Comment:
Any reason to not break this out into a PR now?

Reply:
Without some magic in the compiler the `sink` might turn into a hidden copy if not careful.

I don't want to do the compiler part atm.

Reply:
hmm, sounds like we'll need a `consume` like feature, rather than `sink`